File: Solutions_Easy/0746_MinCostClimbingStairs.py
class Solution:
    def _min_cost_climbing_stairs(self, i: int) -> int:
        if i < 0:
            return 0

        if self.cost_total[i] == -1:
            self.cost_total[i] = min(
                self.cost[i - 1] + self._min_cost_climbing_stairs(i - 1),
                self.cost[i - 2] + self._min_cost_climbing_stairs(i - 2),
            )

        return self.cost_total[i]

    def minCostClimbingStairs(self, cost: List[int]) -> int:
        self.cost = cost
        self.cost_total = [0, 0] + [-1] * (len(cost) - 2)

        n = len(cost)
        self._min_cost_climbing_stairs(n - 1)
        return min(
            self.cost[-1] + self.cost_total[-1], self.cost[-2] + self.cost_total[-2]
        )

File: Solutions_Easy/0746_MinCostClimbingStairs.cpp
class Solution {
private:
    vector<int> cost_total;
    vector<int> cost;

    int _min_cost_climbing_stairs(int i) {
        if (i < 0) {
            return 0;
        }

        if (cost_total[i] == -1) {
            int min_cost = (cost[i - 1] + _min_cost_climbing_stairs(i - 1) < cost[i - 2] + _min_cost_climbing_stairs(i - 2))
                               ? cost[i - 1] + _min_cost_climbing_stairs(i - 1)
                               : cost[i - 2] + _min_cost_climbing_stairs(i - 2);

            cost_total[i] = min_cost;
        }

        return cost_total[i];
    }

public:
    int minCostClimbingStairs(vector<int> &cost) {
        this->cost = cost;
        cost_total = vector<int>(2, 0);     // Initialize the first two elements with 0
        cost_total.resize(cost.size(), -1); // Resize and initialize remaining elements with -1

        int n = cost.size();
        _min_cost_climbing_stairs(n - 1);
        return (cost[n - 1] + cost_total[n - 1] < cost[n - 2] + cost_total[n - 2])
                   ? cost[n - 1] + cost_total[n - 1]
                   : cost[n - 2] + cost_total[n - 2];
    }
};